packages/opencensus-core/src/tags/propagation/binary-serializer.ts

```diff
@@ -0,0 +1,151 @@
+ */
+
+import {TagMap} from '../tag-map';
+import {TagKey, TagValue} from '../types';
+import {DecodeVarint, EncodeVarint} from './variant-encoding';
+
+// This size limit only applies to the bytes representing tag keys and values.
+export const TAG_MAP_SERIALIZED_SIZE_LIMIT = 8192;
+
+const ENCODING = 'utf8';
+const VERSION_ID = 0;
+const TAG_FIELD_ID = 0;
+const VERSION_ID_OFFSET = 0;
+
+/**
+ * Serializes a given TagMap to the on-the-wire format.
+ * @param tagMap The TagMap to serialize.
+ */
+export function serializeBinary(tagMap: TagMap): Buffer {
+  const byteArray: number[] = [];
+  byteArray.push(VERSION_ID);
+  let totalChars = 0;
+  const tags = tagMap.tags;
+  tags.forEach((tagValue: TagValue, tagKey: TagKey) => {
+    totalChars += tagKey.name.length;
+    totalChars += tagValue.value.length;
+    encodeTag(tagKey, tagValue, byteArray);
+  });
+
+  if (totalChars > TAG_MAP_SERIALIZED_SIZE_LIMIT) {
+    throw new Error(`Size of TagMap exceeds the maximum serialized size ${
+        TAG_MAP_SERIALIZED_SIZE_LIMIT}`);
+  }
+  return Buffer.from(byteArray);
+}
+
+/**
+ * Deserializes input to TagMap based on the binary format standard.
+ * @param buffer The TagMap to deserialize.
+ */
+export function deserializeBinary(buffer: Buffer): TagMap {
+  if (buffer.length === 0) {
+    throw new Error('Input buffer can not be empty.');
+  }
+  const versionId = buffer.readInt8(VERSION_ID_OFFSET);
+  if (versionId !== VERSION_ID) {
+    throw new Error(`Wrong Version ID: ${
+        versionId}. Currently supports version up to: ${VERSION_ID}`);
+  }
+  return parseTags(buffer);
+}
+
+function encodeTag(tagKey: TagKey, tagValue: TagValue, byteArray: number[]) {
+  byteArray.push(TAG_FIELD_ID);
+  encodeString(tagKey.name, byteArray);
+  encodeString(tagValue.value, byteArray);
+}
+
+function encodeString(input: string, byteArray: number[]) {
+  byteArray.push(...EncodeVarint(input.length));
+  byteArray.push(...input.split('').map(unicode));
+  return byteArray;
+}
+
+function parseTags(buffer: Buffer): TagMap {
+  const tags = new TagMap();
+  const limit = buffer.length;
+  let totalChars = 0;
+  let currentIndex = 1;
+
+  while (currentIndex < limit) {
+    const fieldId = buffer.readInt8(currentIndex);
+    if (fieldId === TAG_FIELD_ID) {
+      currentIndex += 1;
+      const key = decodeString(buffer, currentIndex);
+      currentIndex += key.length;
+      totalChars += key.length;
+
+      currentIndex += 1;
+      const val = decodeString(buffer, currentIndex);
+      currentIndex += val.length;
+      totalChars += val.length;
+
+      currentIndex += 1;
+      if (totalChars > TAG_MAP_SERIALIZED_SIZE_LIMIT) {
+        throw new Error(`Size of TagMap exceeds the maximum serialized size ${
+            TAG_MAP_SERIALIZED_SIZE_LIMIT}`);
+      } else {
+        tags.set({name: key}, {value: val});
+      }
+    } else {
+      // Stop parsing at the first unknown field ID, since there is no way to
+      // know its length.
+      break;
```

Comment:
What would you think about putting the `break;` condition right at the start of the loop with a `fieldId !== TAG_FIELD_ID` and remove the `else` clause, which would allow the rest of it to be less indented?

Reply:
Sure, done.